File: script/make/protocol/lua/meta/Writer.lua
require("./ProtocolDefine")

Writer = {}

--- This function returns `string`.
--- @param protocol number
--- @param data table
--- @return string
function Writer:write(protocol, data)
    local meta = ProtocolDefine.getWrite(protocol)
    local dataTable = write(meta, 3, data)
    dataTable[1] = ""
    dataTable[2] = ""
    local length = string.len(table.concat(dataTable))
    dataTable[1] = string.pack(">I2", length)
    dataTable[2] = string.pack(">I2", protocol)
    return table.concat(dataTable)
end

function write(metadata, offset, data)
    local dataTable = {}
    for index = 1, #metadata do
        local meta = metadata[index]
        local name = meta["name"]
        local explain = meta["explain"]
        local type = meta["type"]
        if type == "binary" then
            dataTable[offset] = data[name]
        elseif type == "bool" then
            dataTable[offset] = string.pack(">I1", data[name] and 1 or 0)
        elseif type == "u8" then
            dataTable[offset] = string.pack(">I1", data[name])
        elseif type == "u16" then
            dataTable[offset] = string.pack(">I2", data[name])
        elseif type == "u32" then
            dataTable[offset] = string.pack(">I4", data[name])
        elseif type == "u64" then
            dataTable[offset] = string.pack(">I8", data[name])
        elseif type == "i8" then
            dataTable[offset] = string.pack(">i1", data[name])
        elseif type == "i16" then
            dataTable[offset] = string.pack(">i2", data[name])
        elseif type == "i32" then
            dataTable[offset] = string.pack(">i4", data[name])
        elseif type == "i64" then
            dataTable[offset] = string.pack(">i8", data[name])
        elseif type == "f32" then
            dataTable[offset] = string.pack(">f", data[name])
        elseif type == "f64" then
            dataTable[offset] = string.pack(">d", data[name])
        elseif type == "str" then
            dataTable[offset] = string.pack(">s2", data[name])
        elseif type == "bst" then
            dataTable[offset] = string.pack(">s2", data[name])
        elseif type == "rst" then
            dataTable[offset] = string.pack(">s2", data[name])
        elseif type == "list" then
            local listTable = {}
            local listData = data[name]
            listTable[1] = string.pack(">I2", #listData)
            for listIndex = 1, #listData do
                listTable[listIndex + 1] = table.concat(write(explain, 1, listData[listIndex]))
            end
            dataTable[offset] = table.concat(listTable)
        elseif type == "map" then
            local mapTable = {}
            local mapIndex = 1
            local mapData = data[name]
            mapTable[1] = string.pack(">I2", #mapData)
            for _, item in pairs(mapData) do
                mapTable[mapIndex + 1] = table.concat(write(explain, 1, item))
            end
            dataTable[offset] = table.concat(mapTable)
        else
            error(string.format("unknown type: ", meta["type"]))
        end
        offset = offset + 1
    end
    return dataTable
end

File: script/make/protocol/js/meta/Writer.js
import ProtocolDefine from "./ProtocolDefine.js";

export default class Writer {

    constructor() {
        this.textEncoder = new TextEncoder();
    }

    /**
     * write data
     * 
     * @param protocol the protocol number
     * @param data the data object
     * @return ArrayBuffer
     */
    write(protocol, data) {
        // write binary
        const meta = ProtocolDefine.getWrite(protocol);
        let { view, offset } = this.__write__(meta, data, new DataView(new ArrayBuffer(1024)), 4);
        view.setUint16(0, offset - 4, false);
        view.setUint16(2, protocol, false);
        return view.buffer.slice(0, offset);
    }

    __write__(meta, data, view, offset) {
        const type = meta["type"];
        const explain = meta["explain"];
        const key = meta["key"];
        switch (type) {
            case "binary": {
                while (view.byteLength < offset + 6) {
                    let extendView = new DataView(new ArrayBuffer(view.byteLength * 2));
                    (new Uint8Array(extendView.buffer)).set(new Uint8Array(view.buffer));
                    view = extendView;
                }
                // binary
                (new Uint8Array(view.buffer, offset)).set(new Uint8Array(data));
                offset = offset + data.byteLength;
            } break;
            case "bool": {
                while (view.byteLength < offset + 2) {
                    let extendView = new DataView(new ArrayBuffer(view.byteLength * 2));
                    (new Uint8Array(extendView.buffer)).set(new Uint8Array(view.buffer));
                    view = extendView;
                }
                view.setUint8(offset, data ? 1 : 0, false);
                offset = offset + 1;
            } break;
            case "u8": {
                while (view.byteLength < offset + 2) {
                    let extendView = new DataView(new ArrayBuffer(view.byteLength * 2));
                    (new Uint8Array(extendView.buffer)).set(new Uint8Array(view.buffer));
                    view = extendView;
                }
                view.setUint8(offset, data, false);
                offset = offset + 1;
            } break;
            case "u16": {
                while (view.byteLength < offset + 2) {
                    let extendView = new DataView(new ArrayBuffer(view.byteLength * 2));
                    (new Uint8Array(extendView.buffer)).set(new Uint8Array(view.buffer));
                    view = extendView;
                }
                view.setUint16(offset, data, false);
                offset = offset + 2;
            } break;
            case "u32": {
                while (view.byteLength < offset + 4) {
                    let extendView = new DataView(new ArrayBuffer(view.byteLength * 2));
                    (new Uint8Array(extendView.buffer)).set(new Uint8Array(view.buffer));
                    view = extendView;
                }
                view.setUint32(offset, data, false);
                offset = offset + 4;
            } break;
            case "u64": {
                while (view.byteLength < offset + 8) {
                    let extendView = new DataView(new ArrayBuffer(view.byteLength * 2));
                    (new Uint8Array(extendView.buffer)).set(new Uint8Array(view.buffer));
                    view = extendView;
                }
                view.setBigUint64(offset, data, false);
                offset = offset + 8;
            } break;
            case "i8": {
                while (view.byteLength < offset + 8) {
                    let extendView = new DataView(new ArrayBuffer(view.byteLength * 2));
                    (new Uint8Array(extendView.buffer)).set(new Uint8Array(view.buffer));
                    view = extendView;
                }
                view.setInt8(offset, data, false);
                offset = offset + 1;
            } break;
            case "i16": {
                while (view.byteLength < offset + 2) {
                    let extendView = new DataView(new ArrayBuffer(view.byteLength * 2));
                    (new Uint8Array(extendView.buffer)).set(new Uint8Array(view.buffer));
                    view = extendView;
                }
                view.setInt16(offset, data, false);
                offset = offset + 2;
            } break;
            case "i32": {
                while (view.byteLength < offset + 4) {
                    let extendView = new DataView(new ArrayBuffer(view.byteLength * 2));
                    (new Uint8Array(extendView.buffer)).set(new Uint8Array(view.buffer));
                    view = extendView;
                }
                view.setInt32(offset, data, false);
                offset = offset + 4;
            } break;
            case "i64": {
                while (view.byteLength < offset + 8) {
                    let extendView = new DataView(new ArrayBuffer(view.byteLength * 2));
                    (new Uint8Array(extendView.buffer)).set(new Uint8Array(view.buffer));
                    view = extendView;
                }
                view.setBigInt64(offset, data, false);
                offset = offset + 8;
            } break;
            case "f32": {
                while (view.byteLength < offset + 4) {
                    let extendView = new DataView(new ArrayBuffer(view.byteLength * 2));
                    (new Uint8Array(extendView.buffer)).set(new Uint8Array(view.buffer));
                    view = extendView;
                }
                view.setFloat32(offset, data, false);
                offset = offset + 4;
            } break;
            case "f64": {
                while (view.byteLength < offset + 8) {
                    let extendView = new DataView(new ArrayBuffer(view.byteLength * 2));
                    (new Uint8Array(extendView.buffer)).set(new Uint8Array(view.buffer));
                    view = extendView;
                }
                view.setFloat64(offset, data, false);
                offset = offset + 8;
            } break;
            case "bst":
            case "str":
            case "ast": {
                // str
                let strArray = this.textEncoder.encode(data);
                while (view.byteLength < offset + 2) {
                    let extendView = new DataView(new ArrayBuffer(view.byteLength * 2));
                    (new Uint8Array(extendView.buffer)).set(new Uint8Array(view.buffer));
                    view = extendView;
                }
                view.setUint16(offset, strArray.length, false);
                offset = offset + 2;
                // extend
                while (view.byteLength < offset + strArray.length) {
                    let extendView = new DataView(new ArrayBuffer(view.byteLength * 2));
                    (new Uint8Array(extendView.buffer)).set(new Uint8Array(view.buffer));
                    view = extendView;
                }
                (new Uint8Array(view.buffer, offset)).set(strArray);
                offset = offset + strArray.length;
            } break;
            case "list": {
                if(typeof key == 'undefined') {
                    while (view.byteLength < offset + 2) {
                        let extendView = new DataView(new ArrayBuffer(view.byteLength * 2));
                        (new Uint8Array(extendView.buffer)).set(new Uint8Array(view.buffer));
                        view = extendView;
                    }
                    // u16
                    view.setUint16(offset, data.length, false);
                    offset = offset + 2;
                    for (const item of data) {
                        // list only one explain
                        const result = this.__write__(explain[0], item, view, offset);
                        view = result.view;
                        offset = result.offset;
                    }
                } else {
                    while (view.byteLength < offset + 2) {
                        let extendView = new DataView(new ArrayBuffer(view.byteLength * 2));
                        (new Uint8Array(extendView.buffer)).set(new Uint8Array(view.buffer));
                        view = extendView;
                    }
                    // u16
                    view.setUint16(offset, Object.keys(data).length, false);
                    offset = offset + 2;
                    for (const item of Object.values(data)) {
                        // key list only one explain
                        const result = this.__write__(explain[0], item, view, offset);
                        view = result.view;
                        offset = result.offset;
                    }
                }
            } break;
            case "map": {
                for(const sub of explain) {
                    const result = this.__write__(sub, data[sub["name"]], view, offset);
                    view = result.view;
                    offset = result.offset;
                }
            } break;
            default: throw ("unknown type: " + meta["type"]);
        }
        return { view, offset };
    }
}
